Add editor check for compatible visibility layers for `CanvasItem`

AThousandShips opened the pull request:
The wording might need some improvement, and need confirmation that this catches the full behavior of this feature. It doesn't check the cull layers of any viewport, can add if desired but keeping it simple for just the hierarchy (and an error for the viewport might be annoying if you *want* them to be invisible at the start just by the layers, but the hierarchy itself would *usually* be an accident I'd say)

In short: This checks that the condition for visibility is met, i.e. that a `CanvasItem` has at least one layer shared with *all* parents until either the root or one parent that is top-level. To avoid annoying potential false positives it also doesn't check if the layers for the node itself is empty, that feels like something intentional and shouldn't throw a warning

Correctly updates when layers of parents update, as well as when top-level changes for parents and self

* Closes: https://github.com/godotengine/godot/issues/93316

See (I'd say regardless of that one being merged this one is useful, but not vice versa):
* https://github.com/godotengine/godot/pull/93335
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

AThousandShips committed:
Add editor check for compatible visibility layers for `CanvasItem`

AThousandShips commented on scene/gui/control.cpp:
```diff
@@ -248,7 +248,7 @@ void Control::get_argument_options(const StringName &p_function, int p_idx, List
 PackedStringArray Control::get_configuration_warnings() const {
 	ERR_READ_THREAD_GUARD_V(PackedStringArray());
-	PackedStringArray warnings = Node::get_configuration_warnings();
+	PackedStringArray warnings = CanvasItem::get_configuration_warnings();
```
Assuming this is needed, didn't test without it

AThousandShips replied:
Will split this off into its own PR with more fixes like this

kleonc commented on scene/main/canvas_item.cpp:
```diff
@@ -400,6 +400,39 @@ void CanvasItem::_notification(int p_what) {
+	PackedStringArray warnings = Node::get_configuration_warnings();
+
+	// Check compatible visibility layer, only if not top-level and any bit is set.
+	if (!top_level && visibility_layer) {
```
You can simplify this section by using `CanvasItem::get_parent_item`.

kleonc commented on scene/main/canvas_item.cpp:
```diff
@@ -489,6 +522,8 @@ void CanvasItem::set_as_top_level(bool p_top_level) {
 		get_viewport()->canvas_item_top_level_changed();
 	}
+
+	update_configuration_warnings();
```
The child items' warnings might be affected/changed as a result. Does this propagate the update for the children? :thinking:

AThousandShips replied:
It does! Tested out, changing the mask or the top level status of a parent node updates the check of children

kleonc replied:
Hmm, I think it doesn't actually say the children to update their warnings, it's just this single update. So AFAICT technically the answer to my question is no.

But the warnings in the editor's SceneTree dock are updated for the children correctly because it handles the `SceneTree.node_configuration_warning_changed(Node)` signal by updating the whole tree, and thus fetching all warnings anew. So it "just works" nonetheless.

Anyway, I think all / almost all warnings are updated like in here and it's good enough / not an issue because of the above.

AThousandShips replied:
Can move it to the propagation check and see

kleonc replied:
That would be more correct/reliable, yes. But I see that e.g. in CanvasModulate I did it as incorrectly as you in here :upside_down_face:, aka notified only about the one being changed / the one triggering the change, instead of notifying about all which are affected / have their warning changed.

Again, for the in-engine usage it seems it currently doesn't matter, as in the editor the whole scene tree / all Tree items are being recreated anyway. If we'd want to change it to only update relevant items in the Tree then this would became a real issue indeed.

From the user standpoint it might be surprising though. However it seems not reported, aka only a theoretical issue? :thinking:

I'd say let's keep it as is in here, and if we want to correct this then it would make sense to do so in a separate PR by checking/fixing all `update_configuration_warnings()` accross the codebase (no idea how many of them are in a similar manner not notifying about other affected nodes).

AThousandShips replied:
I'll add a note that it relies on the tree being updated for now

AThousandShips replied:
For the top level part there's a propagated method I'll add the check to instead, but I'll add a note to the layer one rather than writing a new propagation for it

kleonc commented on scene/main/canvas_item.cpp:
```diff
@@ -1401,6 +1436,7 @@ void CanvasItem::set_visibility_layer(uint32_t p_visibility_layer) {
 	ERR_THREAD_GUARD;
 	visibility_layer = p_visibility_layer;
 	RenderingServer::get_singleton()->canvas_item_set_visibility_layer(canvas_item, p_visibility_layer);
+	update_configuration_warnings();
```
The child items' warnings might be affected/changed as a result. Does this propagate the update for the children? :thinking:

AThousandShips commented:
There, changed to using `get_parent_item` and moved the configuration update to `_top_level_changed` which is propagated to child nodes to future proof it with that part, and left a note in the layer setter

AThousandShips commented on scene/main/canvas_item.cpp:
```diff
@@ -500,6 +523,7 @@ void CanvasItem::_top_level_changed() {
 			child->_top_level_changed_on_parent();
 		}
 	}
+	update_configuration_warnings();
```
I can move this back to `set_as_top_level` if there would be performance concerns calling it down a deep tree and add a note to this as well, works either way

AThousandShips commented:
Does the visibility check ignore the sub viewport? Is it rendered only based on the root viewport?

JonqsGames commented:
The Visibility is handled per Viewport (as Canvas Cull Mask is a viewport value)
For not CanvasItem node i don't know what the behavior should be, not sure of the policy of the engine on that.

AThousandShips commented:
Well check what it does with rendering and check if it matches this warning, from my reading of the code it matches the actual behaviour, if you have some corrections please add them, otherwise if this matches how things work then nothing needs to be changed right?

AThousandShips commented:
For the viewport cull being ignored I pointed that out in the PR description and explained why

AThousandShips commented:
So if someone on the rendering team could confirm some details of the functioning here if there's something missing, to clarify:
* This seeks to catch cases where the hierarchy is obviously unworkable, where there's little to no reason to assume a setup is valid, where a node can't possibly be visible
* It doesn't care about the flags on the viewport, partially because this isn't easily accessible in the editor, and partially because that could be something the user wants to change to make nodes visible or not later

Here's the assumptions made that I'd like verified (based on the wording in the docs and other conversations as well as reading the code):
* Top level nodes are considered the root for these calculations 
* The check stops with the top most canvas item, either having no canvas item parent, or being top level (based on the server parent assignment code elsewhere), and doesn't jump in the hierarchy or otherwise check for nodes higher

In other words, is there potentially something further needed than the direct traversal of canvas item parents here
